From 363a24c141f67c1d20d2b7e826da3ec5c636dd6e Mon Sep 17 00:00:00 2001
From: Ethan Wei <ethanwei@ami.com>
Date: Sun, 15 Oct 2023 23:03:50 -0700
Subject: [PATCH 05/13] 
 0011-Fix-Dynamic-And-Static-Addrs-Shown-When-IPSrc-Is-DHCP.patch

---
 meson.build                                   |  7 +-
 meson_options.txt                             |  2 +-
 src/ethernet_interface.cpp                    | 26 +++---
 4 files changed, 106 insertions(+), 13 deletions(-)
 create mode 100644 0011-Fix-Dynamic-And-Static-Addrs-Shown-When-IPSrc-Is-DHCP.patch

diff --git a/meson.build b/meson.build
index a8d53b4..db31e99 100644
--- a/meson.build
+++ b/meson.build
@@ -13,7 +13,7 @@ default_busname = 'xyz.openbmc_project.Network'
 conf_data = configuration_data()
 conf_data.set_quoted('DEFAULT_BUSNAME', default_busname)
 conf_data.set('HAVE_UBOOT_ENV', get_option('uboot-env'))
-conf_data.set(
+conf_data.set_quoted(
   'LINK_LOCAL_AUTOCONFIGURATION',
   get_option('default-link-local-autoconf'))
 conf_data.set(
@@ -73,7 +73,10 @@ subdir('src')
 configure_file(
   input: '60-phosphor-networkd-default.network.in',
   output: '60-phosphor-networkd-default.network',
-  configuration: conf_data,
+  configuration: {
+    'LINK_LOCAL_AUTOCONFIGURATION': get_option('default-link-local-autoconf'),
+    'ENABLE_IPV6_ACCEPT_RA': get_option('default-ipv6-accept-ra'),
+  },
   install: true,
   install_dir: dependency('systemd').get_variable(
     'systemdutildir') / 'network')
diff --git a/meson_options.txt b/meson_options.txt
index 56234e6..735335b 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -2,7 +2,7 @@ option('tests', type: 'feature', description: 'Build tests')
 
 option('uboot-env', type: 'boolean',
        description: 'Update u-boot env for MAC changes')
-option('default-link-local-autoconf', type: 'boolean',
+option('default-link-local-autoconf', type: 'string',
        description: 'Enable link local autoconfiguration by default')
 option('default-ipv6-accept-ra', type: 'boolean',
        description: 'Accept router advertisements by default')
diff --git a/src/ethernet_interface.cpp b/src/ethernet_interface.cpp
index d0a4ad4..343a7db 100644
--- a/src/ethernet_interface.cpp
+++ b/src/ethernet_interface.cpp
@@ -700,11 +700,15 @@ void EthernetInterface::writeConfigurationFile()
     {
         auto& network = config.map["Network"].emplace_back();
         auto& lla = network["LinkLocalAddressing"];
-#ifdef LINK_LOCAL_AUTOCONFIGURATION
-        lla.emplace_back("yes");
-#else
-        lla.emplace_back("no");
-#endif
+        if (std::string{LINK_LOCAL_AUTOCONFIGURATION} == "yes")
+            lla.emplace_back("yes");
+        else if (std::string{LINK_LOCAL_AUTOCONFIGURATION} ==  "ipv6")
+            lla.emplace_back("ipv6");
+        else if (std::string{LINK_LOCAL_AUTOCONFIGURATION} ==  "ipv4")
+            lla.emplace_back("ipv4");
+        else
+            lla.emplace_back("no");
+
         network["IPv6AcceptRA"].emplace_back(ipv6AcceptRA() ? "true" : "false");
         network["DHCP"].emplace_back(dhcp4() ? (dhcp6() ? "true" : "ipv4")
                                              : (dhcp6() ? "ipv6" : "false"));
@@ -714,9 +740,13 @@ void EthernetInterface::writeConfigurationFile()
             auto& address = network["Address"];
             for (const auto& addr : addrs)
             {
-                if (originIsManuallyAssigned(addr.second->origin()))
-                {
-                    address.emplace_back(stdplus::toStr(addr.first));
+                if (( addr.second->type() == IP::Protocol::IPv6 && !dhcp6() ) || ( addr.second->type() == IP::Protocol::IPv4 && !dhcp4() )) {
+                    if (originIsManuallyAssigned(addr.second->origin()))
+                    {
+                        address.emplace_back(
+                            fmt::format("{}/{}", addr.second->address(),
+                                        addr.second->prefixLength()));
+                    }
                 }
             }
         }
-- 
2.25.1

